### DISCUSSION TITLE
How to declare special methods and mixins as properties on functional model types

### DISCUSSION THREAD

**Discussion by tadej321** [sequelize/sequelize#15678] [Help & Questions] [answered]
I have a `Questionnaire` model that has a many to many association with the `Section` model:
On the `Questionnaire` model I am trying to use the special methods and mixins such as `createSection` but there is nothing mentioned in the documentation on how to declare these methods for functional models, only for class models. 

I am trying to use the `createSection` method like so:

    import {ContextInterface} from "../../../../../utils/context";
    
    export default async (parent, {uuid}, {models}: ContextInterface): Promise<object> => {
    
      const questionnaire = await models.Questionnaire.findOne({
        where: {
          uuid
        },
        include: { all: true, nested: true }
      })
    
      const newSection = await questionnaire.createSection({section_title: "new section 2"}).catch(e => console.log(e))
    
      return questionnaire
    }

Which is throwing the following typescript error:

> TS2339: Property 'createSection' does not exist on type 'Model '.

On the other hand if I set the type of `questionnaire` constant to `QuestionnaireInstance` like so:

    import {ContextInterface} from "../../../../../utils/context";
    import {QuestionnaireInstance} from "../../../../../database/models/questionnaire";
    
    export default async (parent, {uuid}, {models}: ContextInterface): Promise<object> => {
    
      const questionnaire: QuestionnaireInstance = await models.Questionnaire.findOne({
        where: {
          uuid
        },
        include: { all: true, nested: true }
      })
    
      const newSection = await questionnaire.createSection({section_title: "new section 2"}).catch(e => console.log(e))
    
      return questionnaire
    }

Im getting a different error: 

> Type 'Model ' is missing the following properties from type 'QuestionnaireInstance': createSection, prototype, tableName, primaryKeyAttribute, and 70 more.

Below are the model definitions:

**Questionnaire Model**

    import {
      ModelDefined, Optional
    } from 'sequelize'
    import {Models} from "../index";
    import {SectionInstance} from "./section";

    interface QuestionnaireAttributes {
      id: number;
      uuid: string;
      title: string;
      description: string;
      Sections: SectionInstance[];
      version: number;
      createdAt: Date;
      updatedAt: Date;
    }

    type QuestionnaireCreationAttributes = Optional<QuestionnaireAttributes, 'id' | 'uuid'>;
    
    export interface QuestionnaireInstance extends ModelDefined<QuestionnaireAttributes, QuestionnaireCreationAttributes>, QuestionnaireAttributes {
      associate?(models: Models): void;
      createSection: (values: any, options?: any) => Promise<SectionInstance>;
    }
    
    const questionnaire = (sequelize, DataTypes): QuestionnaireInstance => {
    
      // Define model
      const Questionnaire: QuestionnaireInstance = sequelize.define('Questionnaire', {
        id: {
          type: DataTypes.INTEGER,
          autoIncrement: true,
          primaryKey: true
        },
        uuid: {
          type: DataTypes.UUID,
          defaultValue: DataTypes.UUIDV4,
          allowNull: false,
          unique: true,
          validate: {
            isUUID: 4
          }
        },
        title: DataTypes.STRING,
        description: DataTypes.STRING,
        version: DataTypes.STRING,
      }, {
        sequelize,
        modelName: 'Questionnaire',
        freezeTableName: true
      })
    
      // Define associations
      Questionnaire.associate = models => {
        Questionnaire.belongsToMany(models.Employee, { through: 'Employee_Questionnaire', foreignKey: 'questionnaire_id'})
        Questionnaire.belongsToMany(models.Section, { through: 'Questionnaire_Section', foreignKey: 'questionnaire_id'})
      }
    
    
      // Define hooks
      Questionnaire.beforeCreate(async user => {
    
      });
    
      // Define helper functions
    
      return Questionnaire
    };
    
    export default questionnaire

**Section Model**

    
    import {Optional, ModelDefined} from 'sequelize'
    import {Models} from "../index";
    
    interface SectionAttributes {
      id: number;
      uuid: string;
      section_title: string;
      createdAt: Date;
      updatedAt: Date;
    }
    
    type  SectionCreationAttributes = Optional<SectionAttributes, 'id' | 'uuid'>;
    
    export interface SectionInstance extends ModelDefined<SectionAttributes, SectionCreationAttributes>, SectionAttributes {
      associate?(models: Models): void;
    }
    
    const section = (sequelize, DataTypes) => {
    
      // Define model
      const Section: SectionInstance = sequelize.define('Section', {
        id: {
          type: DataTypes.INTEGER,
          autoIncrement: true,
          primaryKey: true
        },
        uuid: {
          type: DataTypes.UUID,
          defaultValue: DataTypes.UUIDV4,
          allowNull: false,
          unique: true,
          validate: {
            isUUID: 4
          }
        },
        section_title: DataTypes.STRING
      }, {
        sequelize,
        modelName: 'Section',
        freezeTableName: true
      })
    
      // Define associations
      Section.associate = models => {
        Section.belongsToMany(models.Questionnaire, { through: 'Questionnaire_Section', foreignKey: 'section_id'})
        Section.belongsToMany(models.Question, { through: 'Section_Question', foreignKey: 'section_id'})
      }
    
      // Define hooks
      Section.beforeCreate(async section => {
    
      });
    
      // Define helper functions
    
      return Section
    };
    
    export default section

**Comment by ephys** [comment, marked as answer]
Declaring mixin methods & typing your model is described in https://sequelize.org/docs/v6/other-topics/typescript/

**Reply by tadej321** [reply]
It is, but only for class components. Below the class component example is an example of a functional component where there is no explanation on how to do it.

**Reply by ephys** [reply]
You add the definitions to the interface instead of the class

**Reply by tadej321** [reply]
The way I did it above, or is there any other way?

**Reply by ephys** [reply]
The way you define your model does not match what the guide recommends, take a look at https://sequelize.org/docs/v6/other-topics/typescript/#usage-of-sequelizedefine as you may be using an old way of defining things, which is incompatible with what you're trying to do

However I highly recommend using [Sequelize-TypeScript](https://www.npmjs.com/package/sequelize-typescript) for this. Your current approach is going to be very difficult to use in TypeScript

**Reply by tadej321** [reply]
Thanks for the help, I have checked out Sequelize-TypeScript before but I didn’t see any way of implementing Models in a functional way, I guess I will have to go with classes.
